File: vito/drawee-support/src/test/java/com/facebook/fresco/vito/draweesupport/ControllerListenerWrapperTest.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.fresco.vito.draweesupport;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.ArgumentMatchers.isNull;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;
import static org.mockito.Mockito.verifyZeroInteractions;

import android.graphics.drawable.Animatable;
import android.graphics.drawable.Drawable;
import com.facebook.drawee.backends.pipeline.info.ImageOrigin;
import com.facebook.drawee.backends.pipeline.info.ImageOriginListener;
import com.facebook.drawee.controller.ControllerListener;
import com.facebook.imagepipeline.image.ImageInfo;
import org.junit.Before;
import org.junit.Test;

public class ControllerListenerWrapperTest {

  private static final int ID = 123;
  private static final String STRING_ID = "v123";
  private static final String CALLER_CONTEXT = "caller-context";

  private ControllerListener<ImageInfo> mControllerListener;
  private ControllerListenerWrapper mControllerListenerWrapper;

  @Before
  public void setup() {
    mControllerListener = (ControllerListener<ImageInfo>) mock(ControllerListener.class);
    mControllerListenerWrapper = new ControllerListenerWrapper(mControllerListener);
  }

  @Test
  public void testOnSubmit() {
    mControllerListenerWrapper.onSubmit(ID, CALLER_CONTEXT);

    verify(mControllerListener).onSubmit(eq(STRING_ID), eq(CALLER_CONTEXT));
    verifyNoMoreInteractions(mControllerListener);
  }

  @Test
  public void testOnPlaceholderSet() {
    final Drawable placeholder = mock(Drawable.class);
    mControllerListenerWrapper.onPlaceholderSet(ID, placeholder);

    verifyZeroInteractions(mControllerListener);
  }

  @Test
  public void testOnIntermediateImageFailed() {
    final Throwable throwable = new Throwable("Message");
    mControllerListenerWrapper.onIntermediateImageFailed(ID, throwable);

    verify(mControllerListener).onIntermediateImageFailed(eq(STRING_ID), eq(throwable));
    verifyNoMoreInteractions(mControllerListener);
  }

  @Test
  public void testOnFailure() {
    final Throwable throwable = new Throwable("Message");
    final Drawable errorDrawable = mock(Drawable.class);
    mControllerListenerWrapper.onFailure(ID, errorDrawable, throwable);

    verify(mControllerListener).onFailure(eq(STRING_ID), eq(throwable));
    verifyNoMoreInteractions(mControllerListener);
  }

  @Test
  public void testOnRelease() {
    mControllerListenerWrapper.onRelease(ID);

    verify(mControllerListener).onRelease(eq(STRING_ID));
    verifyNoMoreInteractions(mControllerListener);
  }

  @Test
  public void testOnIntermediateImageSet() {
    final ImageInfo imageInfo = mock(ImageInfo.class);
    mControllerListenerWrapper.onIntermediateImageSet(ID, imageInfo);

    verify(mControllerListener).onIntermediateImageSet(eq(STRING_ID), eq(imageInfo));
    verifyNoMoreInteractions(mControllerListener);
  }

  @Test
  public void testOnFinalImageSetNormalDrawable() {
    final Drawable drawable = mock(Drawable.class);
    final ImageInfo imageInfo = mock(ImageInfo.class);
    mControllerListenerWrapper.onFinalImageSet(ID, ImageOrigin.DISK, imageInfo, drawable);

    verify(mControllerListener)
        .onFinalImageSet(eq(STRING_ID), eq(imageInfo), isNull(Animatable.class));
    verifyNoMoreInteractions(mControllerListener);
  }

  @Test
  public void testOnFinalImageSetAnimatedDrawable() {
    final AnimatedDrawable drawable = mock(AnimatedDrawable.class);
    final ImageInfo imageInfo = mock(ImageInfo.class);
    mControllerListenerWrapper.onFinalImageSet(ID, ImageOrigin.DISK, imageInfo, drawable);

    verify(mControllerListener).onFinalImageSet(eq(STRING_ID), eq(imageInfo), eq(drawable));
    verifyNoMoreInteractions(mControllerListener);
  }

  @Test
  public void testOnFinalImageSetImageOriginListener() {
    final ImageOriginListener listener = mock(ImageOriginListener.class);
    final ImageInfo imageInfo = mock(ImageInfo.class);
    ControllerListenerWrapper wrapper =
        ControllerListenerWrapper.create(mControllerListener).setImageOriginListener(listener);

    wrapper.onFinalImageSet(ID, ImageOrigin.DISK, imageInfo, null);

    verify(mControllerListener)
        .onFinalImageSet(eq(STRING_ID), eq(imageInfo), isNull(Animatable.class));
    verify(listener).onImageLoaded(eq(STRING_ID), eq(ImageOrigin.DISK), eq(true), anyString());
    verifyNoMoreInteractions(mControllerListener);
  }

  @Test
  public void testCreateWrapper() {
    ControllerListenerWrapper wrapper = ControllerListenerWrapper.create(mControllerListener);
    assertThat(wrapper).isNotNull();
  }

  @Test
  public void testCreateNullWrapper() {
    ControllerListenerWrapper wrapper = ControllerListenerWrapper.create(null);
    assertThat(wrapper).isNull();
  }

  public abstract static class AnimatedDrawable extends Drawable implements Animatable {}
}

File: vito/drawee-support/src/test/java/com/facebook/fresco/vito/draweesupport/ControllerListenerWrapperTest.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.fresco.vito.draweesupport

import android.graphics.drawable.Animatable
import android.graphics.drawable.Drawable
import com.facebook.drawee.backends.pipeline.info.ImageOrigin
import com.facebook.drawee.controller.ControllerListener
import com.facebook.imagepipeline.image.ImageInfo
import org.assertj.core.api.Assertions.assertThat
import org.junit.Before
import org.junit.Test
import org.mockito.kotlin.eq
import org.mockito.kotlin.isNull
import org.mockito.kotlin.mock
import org.mockito.kotlin.verify
import org.mockito.kotlin.verifyNoMoreInteractions
import org.mockito.kotlin.verifyZeroInteractions

class ControllerListenerWrapperTest {

  private lateinit var controllerListener: ControllerListener<ImageInfo>
  private lateinit var controllerListenerWrapper: ControllerListenerWrapper

  @Before
  fun setup() {
    controllerListener = mock()
    controllerListenerWrapper = ControllerListenerWrapper(controllerListener)
  }

  @Test
  fun testOnSubmit() {
    controllerListenerWrapper.onSubmit(ID.toLong(), CALLER_CONTEXT)
    verify(controllerListener).onSubmit(eq(STRING_ID), eq(CALLER_CONTEXT))
    verifyNoMoreInteractions(controllerListener)
  }

  @Test
  fun testOnPlaceholderSet() {
    val placeholder = mock<Drawable>()
    controllerListenerWrapper.onPlaceholderSet(ID.toLong(), placeholder)
    verifyZeroInteractions(controllerListener)
  }

  @Test
  fun testOnIntermediateImageFailed() {
    val throwable = Throwable("Message")
    controllerListenerWrapper.onIntermediateImageFailed(ID.toLong(), throwable)
    verify(controllerListener).onIntermediateImageFailed(eq(STRING_ID), eq(throwable))
    verifyNoMoreInteractions(controllerListener)
  }

  @Test
  fun testOnFailure() {
    val throwable = Throwable("Message")
    val errorDrawable = mock<Drawable>()
    controllerListenerWrapper.onFailure(ID.toLong(), errorDrawable, throwable)
    verify(controllerListener).onFailure(eq(STRING_ID), eq(throwable))
    verifyNoMoreInteractions(controllerListener)
  }

  @Test
  fun testOnRelease() {
    controllerListenerWrapper.onRelease(ID.toLong())
    verify(controllerListener).onRelease(eq(STRING_ID))
    verifyNoMoreInteractions(controllerListener)
  }

  @Test
  fun testOnIntermediateImageSet() {
    val imageInfo = mock<ImageInfo>()
    controllerListenerWrapper.onIntermediateImageSet(ID.toLong(), imageInfo)
    verify(controllerListener).onIntermediateImageSet(eq(STRING_ID), eq(imageInfo))
    verifyNoMoreInteractions(controllerListener)
  }

  @Test
  fun testOnFinalImageSetNormalDrawable() {
    val drawable = mock<Drawable>()
    val imageInfo = mock<ImageInfo>()
    controllerListenerWrapper.onFinalImageSet(ID.toLong(), ImageOrigin.DISK, imageInfo, drawable)
    verify(controllerListener).onFinalImageSet(eq(STRING_ID), eq(imageInfo), isNull())
    verifyNoMoreInteractions(controllerListener)
  }

  @Test
  fun testOnFinalImageSetAnimatedDrawable() {
    val drawable = mock<AnimatedDrawable>()
    val imageInfo = mock<ImageInfo>()
    controllerListenerWrapper.onFinalImageSet(ID.toLong(), ImageOrigin.DISK, imageInfo, drawable)
    verify(controllerListener).onFinalImageSet(eq(STRING_ID), eq(imageInfo), eq(drawable))
    verifyNoMoreInteractions(controllerListener)
  }

  @Test
  fun testCreateWrapper() {
    val wrapper = ControllerListenerWrapper.create(controllerListener)
    assertThat(wrapper).isNotNull
  }

  @Test
  fun testCreateNullWrapper() {
    val wrapper = ControllerListenerWrapper.create(null)
    assertThat(wrapper).isNull()
  }

  companion object {
    private const val ID = 123
    private const val STRING_ID = "v123"
    private const val CALLER_CONTEXT = "caller-context"
  }

  abstract class AnimatedDrawable : Drawable(), Animatable
}
